[Hack Week] ApiFaker: Home screen and integration with the app

hichamboushaba opened the pull request:
### Description
As part of my Hack Week, I'm bringing back the ApiFaker project paqN3M-SI-p2, and I'm splitting the feature into multiple PRs.

This PR adds the home screen of the ApiFaker module, and integrates with the app (in the developer options screen), the next PR will add the logic to add/edit/delete endpoints.

### Steps to reproduce
1. Use a debug build.
2. Open the settings screen.
3. Tap on Developer Options.
4. Tap on ApiFaker.

### Testing information
- Confirm that the option is visible in Developer Options.
- Confirm that the option opens the ApiFaker home screen.
- Confirm that navigation works as expected.

### The tests that have been performed
^

### Images/gif
<img width=360 src=https://github.com/user-attachments/assets/45e563d8-f621-4115-bcfc-b2575d311ffd/>
<img width=360 src=https://github.com/user-attachments/assets/2c622feb-1417-46c2-a96c-d6ed0eb96040/>

</br>
</br>

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Use hilt viewmodel

hichamboushaba committed:
Use an app's Fragment to allow theming the UI properly

hichamboushaba committed:
Start working on the details screen

hichamboushaba committed:
Handle remaining fields

hichamboushaba committed:
Handle saving changes

hichamboushaba committed:
Fixes to saving and navigation

hichamboushaba committed:
Rename classes

hichamboushaba committed:
Improve the query statement

hichamboushaba committed:
Minor improvements

hichamboushaba committed:
Add a way to enable/disable the API Faker

hichamboushaba committed:
Minor improvements to the home screen's UI

hichamboushaba committed:
Add an initial screen to list the different mocked endpoints

hichamboushaba committed:
Consume the status configuration

hichamboushaba committed:
Add ability to delete requests

hichamboushaba committed:
Add support for filtering by http method

hichamboushaba committed:
Display a hint when ApiFaker is enabled

hichamboushaba committed:
Improve navigation

hichamboushaba committed:
Improve navigation

hichamboushaba committed:
UI improvements

hichamboushaba committed:
Apply internal visibility to some declarations

hichamboushaba committed:
Persist the ApiFaker status

hichamboushaba committed:
Add ability to disable the ApiFaker from the UI hint

hichamboushaba committed:
Add query parameters to the DB model

hichamboushaba committed:
Filter requests using query parameters

hichamboushaba committed:
Add a way to add query parameters in the UI

hichamboushaba committed:
Some fixes

hichamboushaba committed:
Log a warning message if multiple endpoints are matched

hichamboushaba committed:
Add an artifical delay

hichamboushaba committed:
Add the faker library to the app

hichamboushaba committed:
Add an entry to developer options and handle navigation to the API faker

hichamboushaba committed:
Suppress deprecation temporarily

dangermattic commented:
<!--
  6 Errors: Please add tests for class <co...
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          6 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>ApiFakerConfig</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>ApiFakerUiHelper</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>QueryParameterConverter</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>Screen</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EndpointDetailsViewModel</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>HomeViewModel</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Match the TopAppBar background color to the app theme

hichamboushaba committed:
Fix detekt issue

hichamboushaba committed:
Bump db version

hichamboushaba committed:
Handle query parameters extraction for Jetpack tunnel requests

hichamboushaba committed:
Start config flow eagerly

Otherwise we won't get the preference value

hichamboushaba committed:
Fix issue with updating the faker config

hichamboushaba committed:
Set text color as white in the hint

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Remove explicit argument type

Zulu-1540 reviewed (changes requested):
Ok

hichamboushaba committed:
Add some padding between the text fields and their captions

hichamboushaba committed:
Add more tests

JorgeMucientes commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/ui/home/HomeViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+) : ViewModel() {
+    @Suppress("MagicNumber")
+    val endpoints = endpointDao.observeEndpoints()
+        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())
```
Minor np about using the magic number 5000 for this. I understand the reason of it and [where is it coming from](https://medium.com/androiddevelopers/things-to-know-about-flows-sharein-and-statein-operators-20e6ccb2bc74#:~:text=Tip%20for%20Android,used%20in%20ViewModels.): 

> Tip for Android apps! You can use WhileSubscribed(5000) most of the time to keep the upstream flow active for 5 seconds more after the disappearance of the last collector. That avoids restarting the upstream flow in certain situations such as configuration changes. This tip is especially helpful when upstream flows are expensive to create and when these operators are used in ViewModels.

However, when the state flow we are exposing from the viewmodel is a flow we want to keep running until the screen is closed (the host fragment is removed from the stack) why not use `SharingStarted.Lazily` and avoid magic numbers here? Wdyt?

hichamboushaba replied:
If we use `SharingStarted.Lazily` then if the app is background (the user switched to using a different app), then the upstream flow (the DB connection) will stay open and we keep listening to its changes, which is a waste of the device resources.

(But my above statement is true when we use [`collectAsStateWithLifecycle`](https://developer.android.com/reference/kotlin/androidx/lifecycle/compose/package-summary#extension-functions), which I'm not using here now, and which makes `SharingStarted.Lazily` and `SharingStarted.WhileSubscribed(5000)` exactly equivalent 😅. I think the API wasn't stable when I worked on this last year, I'll update it to use it now)

hichamboushaba replied:
This is done now in 9375f2d, please let me know if you still have any concerns with this.

JorgeMucientes replied:
Thanks for updating this to use `collectAdStateWithLifecycle` I wasn't aware of that API. TIL

> If we use SharingStarted.Lazily then if the app is background (the user switched to using a different app), then the upstream flow (the DB connection) will stay open and we keep listening to its changes, which is a waste of the device resources.

While I don't have a strong opinion, specially because we are discussing in the context of a developer feature, I don't think is a big deal to keep the DB connection open while the app is in background and the system is keeping the app resources alive. It might even be handy, because if the DB content changes while the user is looking at another app, when they come back the UI will be fresh right away. Anyway it's really a minor point, for just dealing with another minor thing: the `5000` magic number. 

Let's keep it as is 👍🏼

hichamboushaba replied:
> It might even be handy, because if the DB content changes while the user is looking at another app, when they come back the UI will be fresh right away

The UI will be fresh also when using `SharingStarted.WhileSubscribed(5000)`.

> I don't think is a big deal to keep the DB connection open while the app is in background and the system is keeping the app resources alive..

Yes, I agree it's a very minor point, but since this comes for free with the `collectAdStateWithLifecycle` API, I don't see why we shouldn't have this free optimization 🙂

JorgeMucientes replied:
> but since this comes for free with the collectAdStateWithLifecycle API, I don't see why we shouldn't have this free optimization

Oh sure! My comment about "keeping things as is" was related to keeping `SharingStarted.WhileSubscribed(5000)` vs updating it to use `Lazily`

JorgeMucientes reviewed (approved):
Nice job @hichamboushaba. Everything works as expected. Just left a minor point for discussion but nothing blocking :)

hichamboushaba committed:
Make DB flow collection lifecycle aware

hichamboushaba committed:
Fix UI with long endpoint paths

hichamboushaba committed:
Improve logic of warning log

This makes sure the log is not printed when the matched endpoints use different query parameters.

hichamboushaba committed:
Merge pull request #13083 from woocommerce/task/api-faker-5

[Hack Week] ApiFaker: match endpoints using query parameters

hichamboushaba committed:
Merge pull request #13076 from woocommerce/task/api-faker-4

[Hack Week] ApiFaker: display a hint on the main app UI when the ApiFaker is enabled

hichamboushaba committed:
Merge pull request #13070 from woocommerce/task/api-faker-3

[Hack Week] ApiFaker: Adding, deleting and deleting endpoints
